include/ignition/gazebo/components/ReferenceModels.hh

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+#ifndef IGNITION_GAZEBO_COMPONENTS_REFERENCEMODELS_HH_
+#define IGNITION_GAZEBO_COMPONENTS_REFERENCEMODELS_HH_
+
+#include <iostream>
+#include <set>
+
+#include <ignition/gazebo/Entity.hh>
+#include <ignition/gazebo/components/Factory.hh>
+#include <ignition/gazebo/components/Component.hh>
+#include <ignition/gazebo/config.hh>
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace components
+{
+  /// \brief Data structure that holds information about which models for
+  /// a given link view this link as the model's canonical link. An example
+  /// of multiple models having the same canonical link could be nested models.
+  struct ReferenceModelsInfo
+  {
+    /// \brief A topological ordering of the models which view this link as its
+    /// canonical link.
+    std::set<Entity> models;
```

Comment:
I believe that this is a safe assumption _for now_ because of how models are created in the `SdfEntityCreator` (I made a [comment](https://github.com/ignitionrobotics/ign-gazebo/blob/76cfcf4b4c25919b28278ee4b7d25d217c13bf97/src/systems/physics/Physics.cc#L1820-L1827) about this as a part of my additions to the physics system). If we do add a feature that breaks this assumption down the road, we could always replace the `std::set` of this component with a custom data structure if needed. Maybe I can make the `models` data member `private` and add some more methods to this struct if necessary so that users only interact with this struct via the API? That way, we can change what data structure we use for `models` whenever we need to without breaking code elsewhere (all we need to do is make sure the methods behave appropriately).

Reply:
As a follow-up that just came to mind - if we were to make `models` private and potentially change the data structure of this member later, would that break ABI? If so, then I'll probably need to replace the data structure for `models` with something that is more future-proof before this PR is merged.

Reply:
I believe it would still break ABI. To not break ABI, we've been using the PIMPL pattern, but I'm not a fan of using it for components because it requires heap allocation of the data pointer, which defeats the purpose of the ECM in that the data now resides in a different part of memory, not in the contiguous storage area allocated by the ECM.

Reply:
Agreed :+1: okay, so I think at this point, we have the following options, assuming we go the route of making `models` private and having users interact with this component only through public API:
1. Leave `models` as a `std::set`. This means that if we add any features that break this functionality, they'd need to be added in fortress, and we'd also need to change `std::set` to something else in fortress
2. Change `models` to be something other than `std::set` in order to "future-proof" the possibility of adding features in Edifice that break the assumption made when using `std::set`

Personally, I am a fan of option 1. The reason for this is because if we choose option 2, that would mean more development time on this PR (and more complexity) for no real purpose yet - there's a software engineering principal along the lines of "don't take the time to develop something until you actually need it, since it introduces more complexity now and more maintenance down the road", and in this case, I worry that we would spend additional time making changes for feature(s) that may not even land in Edifice.

Now, if we know for sure that feature(s) will be added to Edifice that require `models` to be something other than `std::set`, then I am fine with making this change now. I'm also open to hearing other options if there are any other ones that you can think of. What do you think?

Reply:
> This still assumes that sorting Entitys in ascending order ensures a topological order. Is that a safe assumption? If we ever add a feature where we can nest an existing model into a new parent model, this assumption will fail.

Follow-up: for now, I believe that entities in ascending order ensures a topological order. If new features are added in the future that violate this assumption, we should be able to change the implementation of canonical link mappings as needed since it's now completely internal to the physics system: https://github.com/ignitionrobotics/ign-gazebo/pull/736#issuecomment-830184043